### ISSUE TITLE
source with multiple sources does not work, sources also down't work

### ISSUE THREAD

**Issue by CaptainSanders** [saltstack/salt#54412]
### Description of Issue
See issue #22747 

It seems this problem had reared its head again. Version 2019.02 (master and minion) no longer accepts source when given multiple sources. Problem is, sources ALSO does not work.

### Setup
 my.cnf:
 file.managed:
   - name: /etc/mysql/my.cnf
     - source:
    - source1
    - source2

Gives this error:

 ID: my.cnf
 Function: file.managed
 Name: /etc/mysql/my.cnf
 Result: False
 Comment: Unable to manage file: none of the specified sources were found

### Steps to Reproduce Issue
Changing the sls to:

 my.cnf:
  file.managed:
   - name: /etc/mysql/my.cnf
     - sources:
       - source1
       - source2

gives the following:

[WARNING ] State for file: /etc/mysql/my.cnf - Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily.

### Versions Report
master:

Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Jan 14 2019, 11:02:34)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5
 
System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-60-generic
         system: Linux
        version: Ubuntu 18.04 bionic

So now neither solution works. When the salt-master is 2018.03 (minion is still 2019.02) the first solution (source:) does work.

**Comment by xeacott** [comment]
Thanks for submitting @CaptainSanders I read through that previous ticket and it seems as though its a regression with using `source` and `sources`. If I'm reading you comment correctly [from here](https://github.com/saltstack/salt/issues/22747), a master on 2018.3 and minion of 2019.2 does allow using just the `source` arg to work correctly?

**Comment by CaptainSanders** [comment]
Yes that is correct. The same minion on an upgraded master is having trouble.

**Comment by CaptainSanders** [comment]
Any news on the progress of this? I gather from the P4 tag that not many people use this, yet personally I'd really like to see this updated.

**Comment by CaptainSanders** [comment]
It seems to be working in 2019.2.2,

**Comment by dahabakuk** [comment]
please update documentation here: https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.managed

when using "sources" instead of "source" it all starts working here as it should.

**Comment by dahabakuk** [comment]
still not working in 2019.2.3. source and sources as described above!

master:

Salt Version:
      Salt: 2019.2.3

Dependency Versions:
           cffi: 1.6.0
       cherrypy: unknown
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.26.3
        libnacl: Not Installed
       M2Crypto: 0.31.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.26.4
         Python: 2.7.5 (default, Jun 11 2019, 14:33:56)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: redhat 7.7 Maipo
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1062.12.1.el7.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 7.7 Maipo

minion:
Salt Version:
           Salt: 2019.2.3

Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.33.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.1
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: Not Installed
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Oct 11 2019, 15:04:54)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 17.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.2
            ZMQ: 4.3.1

System Versions:
           dist: redhat 8.1 Ootpa
         locale: UTF-8
        machine: x86_64
        release: 4.18.0-147.5.1.el8_1.x86_64
         system: Linux
        version: Red Hat Enterprise Linux 8.1 Ootpa

**Comment by dahabakuk** [comment]
ok i found my error and this might actually be a bug in the code:

normally i use:

> - source: salt:///path/too/file/in/salt/root/file.txt

notice the 3 ///

which is found in the fileserver backend "roots" as (using salt-run fileserver.file_list)

> path/too/file/in/salt/root/file.txt

this works without problems be it file.managed or cmd.script etc...

now i want to use multiple sources:

> - source:
>    - salt:///path/too/file/in/salt/root/file.txt-minionspecific
>    - salt:///path/too/file/in/salt/root/file.txt-defaultifnonefound

which results in:

`Unable to manage file: none of the specified sources were found`

ok. lets try how its shown in the file_list listing (according to the source code `parsed_scheme` could also be ''

so using:
> - source:
>    - path/too/file/in/salt/root/file.txt-minionspecific
>    - path/too/file/in/salt/root/file.txt-defaultifnonefound


results in:

`Local file source path/too/file/in/salt/root/file.txt-defaultifnonefound does not exist`

where in reality exactly this file does exist, the `file.txt-minionspecific` does not.

so finally to get this flying like i wanted it to:
> - source:
>    - salt://path/too/file/in/salt/root/file.txt-minionspecific
>    - salt://path/too/file/in/salt/root/file.txt-defaultifnonefound

notice the 2 // (which i never use! see above)

now it just works (tm).

i hope someone with deeper code knowledge can do something with this, i currently have not enough time to dive deeper in the file module code :(

**Comment by dwoz** [comment]
Fix merged into the source tree, pending release. Closing per bulk triage; will ship in the next release cycle. Reopen if this reproduces on a currently supported release once the fix is out.
